### ISSUE TITLE
The plugin will run later. Even though it was declared first.

### ISSUE THREAD

**Issue by hwantage** [ChristianMurphy/postcss-combine-duplicated-selectors#1066]
Thanks to @ChristianMurphy for creating the plugin.

The following error occurred 

```bash
TypeError: node.getIterator is not a function
    at LazyResult.visitTick (D:\Temp\postcss\node_modules\postcss\lib\lazy-result.js:475:33)
    at LazyResult.runAsync (D:\Temp\postcss\node_modules\postcss\lib\lazy-result.js:275:30)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 0)
```
below is my **package.json**
```json
{
  "scripts": {
    "build": "postcss input.css --output output.css"
  },
  "devDependencies": {
    "postcss": "^8.4.38",
    "postcss-clean": "^1.2.2",
    "postcss-cli": "^11.0.0",
    "postcss-combine-duplicated-selectors": "^10.0.3",
    "postcss-delete-duplicate-selector": "^1.2.1"
  }
}
```

below is my **postcss.config.js**
```js
module.exports = {
  plugins: [
	require('postcss-combine-duplicated-selectors'),
	require('postcss-delete-duplicate-selector'),
	require('postcss-clean')
  ]
};
```

I have determined that the error is caused by your plugin and requires a code fix.

Has anyone else encountered the same error?

Removing the `postcss-clean` plugin does not cause the error, but there is a more serious problem.

The `postcss-delete-duplicate-selector` plugin is executed first, even though the `postcss-combine-duplicated-selectors` plugin is declared first.

My Test **input.css** is bellow
```css
.bada {
	color:red;
}

.bada {
	color:blue;
	width:100px;
}
```
**output.css** is below : The `postcss-delete-duplicate-selector` plugin deleted the selector at the bottom.
```css
.bada {
	color:red;
}
```

Here's what I was expecting to see
```css
.bada {
        color:red;
	color:blue;
	width:100px;
}
```

I printed out the console.log and saw that `postcss-combine-duplicated-selectors` was running later.
I changed the prepare() function to a Once() function to make sure it was handled without issue.

If you give me a chance, please allow my merge request. I would like to contribute to your plugin. Otherwise, I would appreciate if you could guide me on how to use it properly.

**Comment by hwantage** [comment]
1.why are you trying to run `postcss-combine-duplicated-selector`s and `postcss-delete-duplicate-selector` together? They seem to offer conflicting ways of achieving the same goal?
=>
The elements of the old and new UI are managed through two CSS files (old.css and new.css).

If the branch source of old.css has been merged and has new selectors, they should be included in new.css as well. 

So I concat the old.css and new.css, and then delete the duplicate selectors below.

My Step is below
1. old.css  =>  `postcss-combine-duplicated-selectors` => old.css
2. new.css => `postcss-combine-duplicated-selectors` => new.css 
3. concat old.css & new.css  => concat.css
4. concat.css => `postcss-delete-duplicate-css` => new.css

**new.css**
```css
.bada {
	/* new.css rules */
	color:red;
	width:100px;
}
```
**old.css**
```css
.bada {
	/* old.css rules */
	color:blue;
}
.added-css {
	/* added rules */
	...
}
```

concat two files
**concat.css**
```css
.bada {
	/* new.css rules */
	color:red;
	width:100px;
}
.bada {
	/* old.css rules */
	color:blue;
}
.added-css {
	/* added rules */
	...
}
```
I used `postcss-delete-duplicate-selector` and my new.css result looks like this
**new.css**
```css
.bada {
	/* new.css rules */
	color:red;
	width:100px;
}
.added-css {
	/* added rules */
	...
}
```
The above process allows for the added-css to be added to both old.css and new.css, and the same element, .bada, that we have in both old.css and new.css, to remain intact in each file.

The `postcss-delete-duplicate-selector` plugin can accomplish what I want by removing duplicate selectors, leaving only the top one.

The `postcss-combine-duplicated-selectors` plugin is doing the work of combining duplicate declarations within old.css and new.css by performing it before concatenating the two files.







2. postcss-delete-duplicate-selector runs first because it forces itself to be first in line by using the imperative API on the first node (the root) https://github.com/hwantage/postcss-delete-duplicate-selector/blob/9bf946eed44de3625e50cfbea7352c6c45eaa6f2/src/index.js#L6-L9, why not move postcss-delete-duplicate-selector to the faster and more standard declarative API? https://postcss.org/api/#plugin (and listening to Rule instead of Once)
=> 
I tried changing the code in the `postcss-delete-duplicate-selector` plugin as you advised.
However, when I use it with `postcss-clean` plugin, I get the same error as when I used your plugin.
I was also able to see the same error when using your plugin with `postcss-clean`.
```bash
TypeError: node.getIterator is not a function
    at LazyResult.visitTick (D:\Temp\postcss\node_modules\postcss\lib\lazy-result.js:475:33)
    at LazyResult.runAsync (D:\Temp\postcss\node_modules\postcss\lib\lazy-result.js:275:30)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Promise.all (index 0)
```

**Comment by ChristianMurphy** [comment]
> old.css => postcss-combine-duplicated-selectors => old.css
> new.css => postcss-combine-duplicated-selectors => new.css
> concat old.css & new.css => concat.css
> concat.css => postcss-delete-duplicate-css => new.css

Okay, this sounds like it's two postcss pipelines.
One for the single file and another for the combined file.
Where do `postcss-combine-duplicated-selectors` and `postcss-delete-duplicate-css` need to run at the same time?

> However, when I use it with postcss-clean plugin, I get the same error as when I used your plugin.

`postcss-clean` has a whole different set of issues, it is an old plugin https://postcss.org/api/#oldplugin.
It doesn't work on the AST at all, it reads the raw CSS and rewrites that https://github.com/leodido/postcss-clean/blob/122c86bc71a9a6aa73fb9bd8499e3f673403f0e6/index.m.js#L6-L25
It will probably break a lot of postcss plugins using the new API.

---

I'd recommend both plugin `postcss-clean` and `postcss-delete-duplicate-css` consider adopting PostCSS's plugin recommended guidelines https://postcss.org/docs/postcss-plugin-guidelines#use-fast-node%E2%80%99s-scanning

**Comment by hwantage** [comment]
It is very important to me that they work in the order declared in postcss.config.js.

I will try to modify the code of `postcss-delete-duplicate-css` according to the guidelines and perform a test.

However, in postcss 8 they can be done simultaneously, so I think I will have to solve the problem without using postcss.config.js.

Your advice made me take a closer look at the documentation of postcss and I gained a lot of understanding and knowledge.

**Comment by ChristianMurphy** [comment]
> It is very important to me that they work in the order declared in postcss.config.js.

I agree.

Based off what I've seen so far, the plugins where issues happen when used with `postcss-combine-duplicated-selectors` don't follow the practices in https://postcss.org/docs/postcss-plugin-guidelines, and the issues are due to those plugins not following the postcss guidelines.

I am happy to add improvements to performance, robustness, and clarity in the code of this plugin. 👍 💯 
I am not interested in doing a workaround in this plugin for a bug in another plugin. 🤷 

> However, in postcss 8 they can be done simultaneously, so I think I will have to solve the problem without using postcss.config.js.

I'm not sure I understand you comment or question, could you rephrase?

> Your advice made me take a closer look at the documentation of postcss and I gained a lot of understanding and knowledge.

There are a lot of great resources in the postcss docs!

**Comment by ChristianMurphy** [comment]
> It is very important to me that they work in the order declared in postcss.config.js.

That's not really how Postcss 8 works.
It does a single pass across a file, not multiple ordered passes.

---

Based off what I've seen so far, the plugins where issues happen when used with `postcss-combine-duplicated-selectors` don't follow the practices in https://postcss.org/docs/postcss-plugin-guidelines, and the issues are due to those plugins not following the postcss guidelines.

I am happy to add improvements to performance, robustness, and clarity in the code of this plugin. 👍 💯 
I am not interested in doing a workaround in this plugin for a bug in another plugin. 🤷 

> However, in postcss 8 they can be done simultaneously, so I think I will have to solve the problem without using postcss.config.js.

I'm not sure I understand you comment or question, could you rephrase?

> Your advice made me take a closer look at the documentation of postcss and I gained a lot of understanding and knowledge.

There are a lot of great resources in the postcss docs!

**Comment by hwantage** [comment]
> However, in postcss 8 they can be done simultaneously, so I think I will have to solve the problem without using postcss.config.js.

If I write the plugins in the postcss.config.js file in the order of execution, they may behave differently than I expect. 
This is due to internal behavior that tries to improve the performance of postcss by running plugins in parallel.

For this reason, I recommend running plugins individually using the CLI.
Is there a better way?

In conclusion, your plugin is fine and I appreciate your diligence in responding to my ignorant questions.

**Comment by ChristianMurphy** [comment]
> For this reason, I recommend running plugins individually using the CLI.

That works!

> Is there a better way?

Maybe?

Based off
 
>    old.css => postcss-combine-duplicated-selectors => old.css
>    new.css => postcss-combine-duplicated-selectors => new.css
>    concat old.css & new.css => concat.css
>    concat.css => postcss-delete-duplicate-css => new.css

You may want to use a build tool like rollup or vite to help combine the files together and run everything in one build pipeline.
